### PR TITLE
graph pass dI/dW split example

### PR DESCRIPTION
**Pull request by bdhirsh**
Adding a graph pass to split up the backward graph into subgraphs for computing input gradients and weight gradients. I made my changes off of Sanket's PR here: https://github.com/meta-pytorch/autoparallel/pull/205 (we can also just abandon this PR and fold them in as necessary).

In the example, here are the relevant backward graphs:
* full backward: P2012417569
* bw inputs only: P2012419228
* bw weights only: P2012420054

I ended up implementing the partitioning by just re-using the default partitioner. We could also consider writing our own one-off pass, but the default partitioner seemed to mostly work out of the box. The two changes I had to make were:

(1) removing AC recompute metadata from the backward graph, so the default partitioner wouldn't think we are trying to run AC. This metadata shouldn't matter for now, although I could maybe imagine a version of the dI/dW subgraph splitting where there are intermediates that we compute during dI, that we *want* to recompute in the dW subgraph.

(2) flipping the order of the outputs of the backward graph, so it is (*grad_inputs, *grad_weights). This made it easier to re-use the default partitioner, which expects that the "first" graph that gets split outs maps to the first K outputs of the original graph.

### CODE REVIEW

File: autoparallel/_passes/graph_multiplex.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
```
Comment (bdhirsh) [inline autoparallel/_passes/graph_multiplex.py]: The only real changes in this PR are `example_llama3_di_dw.py` and `split_di_dw_graph.py`, the rest is an artifact of me working on top of https://github.com/meta-pytorch/autoparallel/pull/205 and making the PR against main. I can move this graph pass somewhere else and/or wait for that other PR to land as necessary

File: examples/example_llama3_di_dw.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    #   so I can generate separate dI and dW gradients to carve out subgraphs for
+    # - The input to tok_embeddings is an integral tensor which can't require grad.
+    # Another option would be to manually apply autoparallel to a single transformer block layer.
+    m.tok_embeddings = None
+    return m
+
+
+def input_fn():
+    # 8192 for 70B
+    x = torch.randn(batch_size, seqlen, 4096, device=device, requires_grad=True)
```
Comment (bdhirsh) [inline examples/example_llama3_di_dw.py]: in my example i need the input to the forward/backward region we are tracing to require grad, so I can properly get a backward graph that computes input gradients. In the real pipeline example, this will probably work by us using the pipeline graph split APIs to get a subgraph for each model layer, where its input should already require grad.

File: autoparallel/_passes/split_di_dw_graph.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    new_out_tuple = grad_inputs + grad_weights
+    with bw_gm.graph.inserting_after(output):
+        # TODO: also set the new node's meta properly
+        new_out = bw_gm.graph.output(new_out_tuple)
+    output.replace_all_uses_with(new_out)
+    bw_gm.graph.erase_node(output)
+    return len(grad_inputs)
+
+# TODO: in theory we can infer num_weight_gradients from the graph metadata directly
+def split_di_dw_graph(bw_gm: fx.GraphModule, *, num_weight_gradients) -> tuple[fx.GraphModule, fx.GraphModule]:
```
Comment (ezyang) [inline autoparallel/_passes/split_di_dw_graph.py]: Help me understand what bw_gm is; is this ONLY the linear?

File: examples/example_llama3_di_dw.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+        print(main_b_gm.graph)
+        print("Prefetch Bwd Graph:")
+        print(prefetch_b_gm.graph)
+        multiplexed_gm = multiplex_fw_bw_graph(main_f_gm, main_b_gm)
+        print("Multiplexed Graph:")
+        print(multiplexed_gm.graph)
+        # in the AOTAutograd bw graph, the first K outputs correspond
+        # to gradients for any params/buffers in the user model
+        num_weight_gradients = autop.joint_with_descriptors._aot_state.aot_config.num_params_buffers
+        main_b_gm_di, main_b_gm_dw = split_di_dw_graph(main_b_gm, num_weight_gradients=num_weight_gradients)
```
Comment (ezyang) [inline examples/example_llama3_di_dw.py]: nope it's the full main backward graph

**Review by fmassa — APPROVED** [review]
This LGTM, thanks @bdhirsh !

I'm ok landing the PR as is, but I think it would be good to convert the example into a test.

Also, I'll let you decide wrt Ed's comment on descriptors, might be better to use it indeed in the long run
